### PR TITLE
MultiFab & Array4

### PR DESCRIPTION
**Pull request by ax3l**
Reimplementation of the successful MFiter concepts from: https://github.com/sayerhs/pyamrex for `MultiFab`, avoiding copies. Reimplementation of `yield` continuations with an iterator.

Start new concepts for `Array4` - first via `__array_interface__`:  #17 / https://github.com/AMReX-Codes/pyamrex/issues/9#issuecomment-778679635

- we will make Array4(Real/Int) createable only by MultiFabs (`amrex::Box` indexing semantics by default for everything: index `-1` is ghost)
- add explicit `asarray` / `as_numpy` / `as_cupy` functions for non-owning data from it: (`numpy`/Python Buffer Interface semantics for everything: index `-1` is end of array)
- make data copies explicit or fail if coming from device (check also pytorch & tensorflow interfaces for good vibes)

Refs.:
- https://github.com/sayerhs/pyamrex/blob/master/amrex/amrex_base.pyx#L502-L511
- https://github.com/sayerhs/pyamrex/blob/master/amrex/cpp/multifab.pxi
- https://sayerhs.github.io/pyamrex/index.html#amrex.amrex_base.MultiFab
```py
# assume a MultiFab mfab with 3 components
ngv = mfab.nGrowVect

# Looping
for mfi in mfab:
    bx = mfi.tilebox().grow(ngv)
    marr = mfab.array(mfi) # Array4: added __array_interface__ here

    # index by index assignment
    # - this is AMReX Array4, F-order indices
    # - even though we iterate by fastest varying index, such loops are naturally very slow in Python
    for i, j, k in bx:
        mar[i, j, k, 0] = 10.0 * i
        mar[i, j, k, 1] = 10.0 * j
        mar[i, j, k, 2] = 10.0 * k

    # numpy representation: non-copying view, including the
    # guard/ghost region
    #   note: in numpy, indices are in C-order!
    marr_np = np.array(marr, copy=False)

    # all components at all indices set at once to 42
    marr_np[:, :, :, :] = 42.
    # first component at all indices set at once to 12
    marr_np[0, :, :, :] = 12.

    # check the values at start/end are the same between MultiFab and numpy view in it: first component
    assert(marr_np[0, 0, 0, 0] == marr[bx.small_end])
    assert(marr_np[0, -1, -1, -1] == marr[bx.big_end])
```

### CODE REVIEW

**Comment by ax3l** [comment]
A comment by @WeiqunZhang
> MultiFab contains a vector of ints.  These ints are global box index that can be used to access Fab, MulitiFab::operator[](int K).  There is a public function that returns the vector.  Maybe this can be used to build an iterator in python without involving MFIter.
> FabArrayBase::indexArray() if I recall it correctly.

File: src/Base/Box.cpp [hunk 1]
```diff
@@ -4,19 +4,72 @@
...
+            m_it = m_box.smallEnd();
+        }
+
+        Box3DConstIter& operator++() {
+            // from FABio_ascii::write
+            if (m_it < m_box.bigEnd()) {
+                m_box.next(m_it.value());
+                return *this;
+            }
+            else
```
Comment (ax3l) [inline src/Base/Box.cpp:39]: Maybe `==` and `else` is cheaper & clearer here.

**Comment by ax3l** [comment]
Windows memory violation: Cannot find any issue on Linux with address + undefined sanitizers so far... (besides #29)

update: I used a too-small integer type to store the pointer: https://github.com/numpy/numpy/pull/21248

**Comment by ax3l** [comment]
Array interface on Ubuntu:
```
{'data': (106652627910208, False), 'strides': None, 'descr': [('', '<f8')], 'typestr': '<f8', 'shape': (2, 3, 4), 'version': 3}
```
```
{'data': (106652627910208, False), 'typestr': 'd', 'shape': (2, 3, 4, 1), 'strides': (96, 32, 8, 32), 'version': 3}
```
maybe the 4th dim stride is off after all.

File: tests/test_array4.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+def test_array4(capsys):
+    # from numpy (also a non-owning view)
+    x = np.ones((2, 3, 4,))
+    with capsys.disabled():
+        print(x.__array_interface__)
+    arr = amrex.Array4_double(x)
+    with capsys.disabled():
+        print(arr.__array_interface__)
+    assert(arr.nComp == 1)
+    return
```
Comment (ax3l) [inline tests/test_array4.py]: hack: remove

**Comment by ax3l** [comment]
## Windows
### MSVC
```
{'data': (2382321214176, False), 'strides': None, 'descr': [('', '<f8')], 'typestr': '<f8', 'shape': (2, 3, 4), 'version': 3}
{'data': (-1385635104, False), 'typestr': 'd', 'shape': (2, 3, 4, 1), 'strides': (96, 32, 8, 32), 'version': 3}
```
### Clang
```
{'data': (2255367229088, False), 'strides': None, 'descr': [('', '<f8')], 'typestr': '<f8', 'shape': (2, 3, 4), 'version': 3}
{'data': (509398688, False), 'typestr': 'd', 'shape': (2, 3, 4, 1), 'strides': (96, 32, 8, 32), 'version': 3}
```

**Comment by ax3l** [comment]
## Windows now (failing)
### MSVC
```
tests/test_array4.py::test_array4
(int)buf.strides.at(0)=96
(int)buf.strides.at(1)=32
(int)buf.strides.at(2)=8
{'data': (2430664277264, False), 'strides': None, 'descr': [('', '<f8')], 'typestr': '<f8', 'shape': (2, 3, 4), 'version': 3}
float64
{'data': (-287212272, False), 'shape': (1, 2, 3, 4), 'strides': (192, 96, 32, 8), 'typestr': 'd', 'version': 3}
```
### Clang
```
tests/test_array4.py::test_array4
(int)buf.strides.at(0)=96
(int)buf.strides.at(1)=32
(int)buf.strides.at(2)=8
{'data': (2508376377952, False), 'strides': None, 'descr': [('', '<f8')], 'typestr': '<f8', 'shape': (2, 3, 4), 'version': 3}
float64
{'data': (115477088, False), 'shape': (1, 2, 3, 4), 'strides': (192, 96, 32, 8), 'typestr': 'd', 'version': 3}
```

## Ubuntu now (passing)
```
tests/test_array4.py::test_array4
{'data': (106652627910208, False), 'strides': None, 'descr': [('', '<f8')], 'typestr': '<f8', 'shape': (2, 3, 4), 'version': 3}
float64
(int)buf.strides.at(0)=96
(int)buf.strides.at(1)=32
(int)buf.strides.at(2)=8
{'data': (106652627910208, False), 'shape': (1, 2, 3, 4), 'strides': (192, 96, 32, 8), 'typestr': 'd', 'version': 3}
```
